src/aeso_stdlib.erl

```diff
@@ -0,0 +1,199 @@
+%%%-------------------------------------------------------------------
+%%% @author Radosław Rowicki
+%%% @copyright (C) 2019, Aeternity Anstalt
+%%% @doc
+%%%     Standard library for Sophia
+%%% @end
+%%% Created : 6 July 2019
+%%%
+%%%-------------------------------------------------------------------
+
+-module(aeso_stdlib).
+
+-export([stdlib/0]).
+
+stdlib() ->
+    maps:from_list(
+      [ {<<"List.aes">>, std_list()}
+     %% , {<<"Func.aes">>, std_function()}
+      ]
+     ).
+
+std_function() ->
+"
+namespace Func =
+
+  function id(x) = x
+
+  function const(x) = (y) => x
+
+  function comp(f, g) = (x) => f(g(x))
+
+  function iter(n, f) = iter_(n, f, (x) => x)
+  private function iter_(n, f, acc) =
+    if(n == 0) acc
+    elif(n == 1) comp(f, acc)
+    else iter_(n / 2, f, if(n mod 2 == 0) comp(acc, acc) else comp(f, comp(acc, acc)))
+
+  function curry2(f) = (x) => (y) => f(x, y)
+  function curry3(f) = (x) => (y) => (z) => f(x, y, z)
+
+  function uncurry2(f) = (x, y) => f(x)(y)
+  function uncurry3(f) = (x, y, z) => f(x)(y)(z)
+
+  function tuplify2(f) = (t) => switch(t)
+    (x, y) => f(x, y)
+//  function tuplify3(f) = (t) => switch(t)
+//    (x, y, z) => f(x, y, z)
+
+  function untuplify2(f) = (x, y) => f((x, y))
+  function untuplify3(f) = (x, y, z) => f((x, y, z))
+".
+
+std_list() ->
+"
+namespace List =
+
+  function empty(l) = switch(l)
+    [] => true
+    _  => false
+
+
+  function first(l) = switch(l)
+    [] => None
+    h::_ => Some(h)
+
+  function tail(l) = switch(l)
+    [] => None
+    _::t => Some(t)
+
+  function last(l) = switch(l)
+    [] => None
+    [x] => Some(x)
+    _::t => last(t)
+
+  function find(p, l) = switch(l)
+    [] => None
+    h::t => if(p(h)) Some(h) else find(p, t)
+
+  function nth(n, l) = switch(l)
+    h::t => if(n == 0) Some(h) else nth(n-1, t)
+    [] => None
```

Comment:
To make these `option`-valued functions convenient to use, we should should also have some standard functions on options (but let's leave that for another PR)

Reply:
That's good idea. Maybe we should start an issue for it?

Reply:
I have added some support